hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/SCMContainerManager.java

```diff
@@ -101,12 +103,23 @@ public SCMContainerManager(
     this.numContainerPerVolume = conf
         .getInt(ScmConfigKeys.OZONE_SCM_PIPELINE_OWNER_CONTAINER_COUNT,
             ScmConfigKeys.OZONE_SCM_PIPELINE_OWNER_CONTAINER_COUNT_DEFAULT);
+    this.numPipelinesPerRaftLogDisk = conf
+        .getInt(ScmConfigKeys.OZONE_SCM_PIPELINE_PER_RAFT_LOG_DISK,
+            ScmConfigKeys.OZONE_SCM_PIPELINE_PER_RAFT_LOG_DISK_DEFAULT);
 
     loadExistingContainers();
 
     scmContainerManagerMetrics = SCMContainerManagerMetrics.create();
   }
 
+  private int getOpenContainerCountPerPipeline(Pipeline pipeline) {
+    int totalContainerCountPerDn = numContainerPerVolume *
+        pipelineManager.getNumHealthyVolumes(pipeline);
+    int maxPipelineCountPerDn = pipelineManager.maxPipelineLimit(pipeline);
+    return (int) Math.ceil(
+        ((double) totalContainerCountPerDn / maxPipelineCountPerDn));
+  }
+
```

Comment:
Will this work in case of heterogeneous datanodes, where one datanode has 1 Raft log disk with 2 data disk and the other datanode has 5 Raft log disk with 10 data disk?

According to the current logic `getOpenContainerCountPerPipeline` will return 10, if `numContainerPerVolume` and `numPipelinesPerRaftLogDisk` are set to 2.

Reply:
might need change pipeline placement policy to make sure that only allocating pipelines on homogeneous machiens.

It one pipeline connects a strong DN and weak DN, its open container number will be `volumeNumOfStrongDN * 3 / pipelineNumOfWeakDN.`.

Reply:
I agree with @GlenGeng here. The pipeline placement should choose homogeneous datanodes. The choice being made here, have many containers open on a minimal set of pipelines that we can have out of the set of datanodes in the pipleines.

Reply:
Changed the logic to take the min of the data disks count of the set of dns in a pipeline while determining the open container limit.